docs/source/training.rst

```diff
@@ -10,290 +10,362 @@
-a :func:`~transformers.BatchEncoding` instance which prepares everything we might need to pass to the model.
+First, let's define our model:
 
 .. code-block:: python
 
-    from transformers import BertTokenizer
-    tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
-    text_batch = ["I love Pixar.", "I don't care for Pixar."]
-    encoding = tokenizer(text_batch, return_tensors='pt', padding=True, truncation=True)
-    input_ids = encoding['input_ids']
-    attention_mask = encoding['attention_mask']
+    from transformers import AutoModelForSequenceClassification
 
-When we call a classification model with the ``labels`` argument, the first returned element is the Cross Entropy loss
-between the predictions and the passed labels. Having already set up our optimizer, we can then do a backwards pass and
-update the weights:
+    model = AutoModelForSequenceClassification.from_pretrained("bert-base-cased", num_labels=2)
+
+Then, to define our :class:`~transformers.Trainer`, we will need to instantiate a
+:class:`~transformers.TrainingArguments`. This class contains all the hyperparameters we can tune for the
+:class:`~transformers.Trainer` or the flags to activate the different training options it supports. Let's begin by
+using all the defaults, the only thing we then have to provide is a directory in which the checkpoints will be saved:
 
 .. code-block:: python
 
-    labels = torch.tensor([1,0]).unsqueeze(0)
-    outputs = model(input_ids, attention_mask=attention_mask, labels=labels)
-    loss = outputs.loss
-    loss.backward()
-    optimizer.step()
+    from transformers import TrainingArguments
+
+    training_args = TrainingArguments("test_trainer")
 
-Alternatively, you can just get the logits and calculate the loss yourself. The following is equivalent to the previous
-example:
+Then we can instantiate a :class:`~transformers.Trainer` like this:
 
 .. code-block:: python
 
-    from torch.nn import functional as F
-    labels = torch.tensor([1,0])
-    outputs = model(input_ids, attention_mask=attention_mask)
-    loss = F.cross_entropy(outputs.logits, labels)
-    loss.backward()
-    optimizer.step()
+    from transformers import Trainer
 
-Of course, you can train on GPU by calling ``to('cuda')`` on the model and inputs as usual.
+    trainer = Trainer(
+        model=model, args=training_args, train_dataset=small_train_dataset, eval_dataset=small_eval_dataset
+    )
 
-We also provide a few learning rate scheduling tools. With the following, we can set up a scheduler which warms up for
-``num_warmup_steps`` and then linearly decays to 0 by the end of training.
+To fine-tune our model, we just need to call
 
 .. code-block:: python
 
-    from transformers import get_linear_schedule_with_warmup
-    scheduler = get_linear_schedule_with_warmup(optimizer, num_warmup_steps, num_train_steps)
+    trainer.train()
```

Comment:
Some question that could be worth including in this tutorial:

- Does `Trainer` automatically freeze the pre-trained base and unfreeze the head for you by default?  
- Is `Trainer` able to somehow unfreeze the entire model if you like with some kind of parameter?
- If you modify the `requires_grad` parameter of the underlying model does Trainer respect that?
- Is `Trainer` only for PyTorch models?

Reply:
There is no freezing baked into the `Trainer` as it's not useful for Transformers models (or if it is, no one has found the right way to do it yet). If you pass a model with some parameters with no requires grad, the Trainer will respect that. It does not touch the model apart from putting it eval/training mode.

Reply:
Wow ok, I didn't know that!  I guess its somewhat surprising that you can fine-tune the whole thing at once even if the head is completely new / randomly initialized.  For some reason, I thought that folks freeze the base and train just the head, and then afterward unfreeze the whole thing -> but maybe that's just some habit I have from non-transformer models?  I guess I have to unlearn a couple of things :)   That's pretty fascinating I will try to read about why this works!  (If you know of anything that explains this please LMK as I would love to learn)

Reply:
I don't think there are articles studying why it works, on my side it's mostly though various experiments I've noticed this.